helm-v3/005-test.md

```diff
@@ -0,0 +1,44 @@
+# Test Command
+
+## Overview
+
+`helm test` gives a chart author the ability to define actions on a helm release which validate that the chart is running as intended. This is useful for ensuring that a `helm install` or `helm upgrade` created/modified Kubernetes resources successfully and that the consequential `helm release` is functional.
+* A **test** is defined as a command and some arguments executed from a container running in a Kubernetes Pod. If the command completes inside the container with an exit code of 0, it is successful. All other exit codes indicate the test failed.
+* Tests must be defined in the `templates/tests/` directory. This used to be recommended as a best practice. See motivations for change below:
+  * Makes it much easier for a human to identify which templates are tests vs. which ones are manifests that define the application
+  * There is currently no built-in way to create supporting resources to use for running a test. Having a test directory will allow us to more easily have a way to organize and manage resources directly related to testing.
+  * This makes initial rendering and filtering resource manifests in the rendering engine much simpler to implement.
+However, making this a requirement will make the logic for rendering resources much easier to
+* Test annotations
+  * Each test pod must have an annotation to denote the expected condition.
+    * `helm.sh/test-expect-success: <true/false>`
+      * `helm.sh/test-expect-success: true` is used to 1) denote that the pod the annotation is applied to is a test and not a resource to be created during `helm install` and 2) denote the expected outcome of the test.
```

Comment:
I'd like to eliminate failure test cases. In practice they are hardly used -- on github it appears that there are about [20 usages of test-failure](https://github.com/search?l=YAML&q=helm.sh%2Fhook+test-failure&type=Code) and [8000 usages of test-success](https://github.com/search?l=YAML&q=helm.sh%2Fhook+test-success&type=Code) (0 and 34 respectively in helm/charts).

Where `test-failure` _is_ used it can be replaced by a small amount of logic in the test itself (and this is a better fit since there are a variety of ways a pod can fail aside from what runs in its entrypoint).